Add rki-covid integration

thebino opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
RKI Covid numbers is an integration to visualize covid numbers of german districts. Data provided by the official responsible "Robert-Koch Institut"


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->
First draft of the integration (previous published via HACS)

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [x] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

thebino committed:
Add rki-covid integration

thebino committed:
Add rki-covid integration

thebino committed:
Add hospitalization to rki_covid integration

thebino committed:
Add hospitalization to rki_covid integration

bachya commented:
Should this perhaps be included in the existing COVID integration? https://www.home-assistant.io/integrations/coronavirus

thebino commented:
@bachya 
I don't think so, the `coronavirus` integration is global and on state level.
This one is only for germany and on distric level.

gjohansson-ST commented on homeassistant/components/rki_covid/__init__.py:
```diff
@@ -0,0 +1,172 @@
+from .data import DistrictData, StateData
+
+_LOGGER = logging.getLogger(__name__)
+PLATFORMS = ["sensor"]
```
Use Platform enum https://github.com/home-assistant/core/blob/dev/homeassistant/const.py#L22

thebino replied:
done

iMicknl replied:
It seems you reverted this change?

gjohansson-ST replied:
He never committed it

thebino replied:
I did, but had to revert it due to an error while running locally.
[9f2d1a2cb4dcff7dac7d0c6bd699afd46ba1ebf6](https://github.com/home-assistant/core/pull/60506/commits/9f2d1a2cb4dcff7dac7d0c6bd699afd46ba1ebf6)

But now I changed it again now. Sorry for the confusion

gjohansson-ST commented on homeassistant/components/rki_covid/__init__.py:
```diff
@@ -0,0 +1,172 @@
+PLATFORMS = ["sensor"]
+
+
+async def async_setup(hass: HomeAssistant, config: dict) -> bool:
+    """Set up the component into HomeAssistant."""
+    _LOGGER.debug("setup component.")
+    parser = RkiCovidParser(aiohttp_client.async_get_clientsession(hass))
+
+    # Make sure coordinator is initialized.
+    await get_coordinator(hass, parser)
+
+    # Return boolean to indicate that initialization was successful.
+    return True
```
Can be removed since loading from config entry

thebino replied:
The whole block or just the return?

gjohansson-ST replied:
The whole block can go as we it should only be used from a config entry

thebino replied:
done

gjohansson-ST commented on homeassistant/components/rki_covid/__init__.py:
```diff
@@ -0,0 +1,172 @@
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up component from a config entry."""
+    _LOGGER.debug(f"Setup item from config entry: {entry.data}.")
+    # Forward the setup to the sensor platform.
```
```suggestion
```

thebino replied:
Since it is a verbose|debug log, it should not matter to stay inside to add logging if necessary

gjohansson-ST replied:
True but for me a quite pointless debugging in production code

gjohansson-ST commented on homeassistant/components/rki_covid/__init__.py:
```diff
@@ -0,0 +1,172 @@
+    """Set up component from a config entry."""
+    _LOGGER.debug(f"Setup item from config entry: {entry.data}.")
+    # Forward the setup to the sensor platform.
+    for component in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, component)
+        )
```
```suggestion
    hass.config_entries.async_setup_platforms(entry, PLATFORMS)
```

gjohansson-ST commented on homeassistant/components/rki_covid/__init__.py:
```diff
@@ -0,0 +1,172 @@
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    _LOGGER.debug(f"Unload item from config entry: {entry.data}.")
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, component)
+                for component in PLATFORMS
+            ]
+        )
+    )
+
+    return unload_ok
```
```suggestion
    return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
```

gjohansson-ST commented on homeassistant/components/rki_covid/config_flow.py:
```diff
@@ -0,0 +1,62 @@
+    """RKI Covid numbers config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_CLOUD_POLL
```
```suggestion
```
Not used anymore

gjohansson-ST commented on homeassistant/components/rki_covid/config_flow.py:
```diff
@@ -0,0 +1,62 @@
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_CLOUD_POLL
+
+    _options: Optional[Dict[str, Any]] = None
```
```suggestion
    _options: dict[str, Any] | None = None
```
use `from __future__ import annotations`

gjohansson-ST replied:
And don't seem to be needed

thebino replied:
> Optional[X] is equivalent to X | None
https://docs.python.org/3/library/typing.html#typing.Optional

gjohansson-ST replied:
I know but it's more clearer to write as such which is asked

thebino replied:
I'll keep this for now. The core is still using python 3.9

iMicknl replied:
I think that core preference is to use the option suggested by @gjohansson-ST, while importing `from __future__ import annotations`.

thebino replied:
adding the annotations import solved the warning. Thanks

gjohansson-ST commented on homeassistant/components/rki_covid/sensor.py:
```diff
@@ -0,0 +1,207 @@
+}
+
+
+async def async_setup_platform(
+    hass: HomeAssistantType,
+    config: ConfigType,
+    async_add_entities: Callable,
+    discovery_info: Optional[DiscoveryInfoType] = None,
+) -> None:
+    """Set up the sensor platform."""
+    _LOGGER.debug("setup sensor for platform")
+    session = async_get_clientsession(hass)
+
+    if CONF_BASEURL in config:
+        _LOGGER.warning("Baseurl is not supported anymore.")
+
+    parser = RkiCovidParser(session)
+    coordinator = await get_coordinator(hass, parser)
+
+    if coordinator is None or coordinator.data is None:
+        raise PlatformNotReady("Data coordinator could not be initialized!")
+
+    districts = config[CONF_DISTRICTS]
+
+    sensors = [
+        RKICovidNumbersSensor(coordinator, district[CONF_DISTRICT_NAME], info_type)
+        for info_type in SENSORS
+        for district in districts
+    ]
+    async_add_entities(sensors, update_before_add=True)
```
Remove as we should load the integration from config entry

thebino replied:
I recall the integration guide recommends adding a configuration setup as well. But I might be wrong here.

gjohansson-ST replied:
Don't think it's specifically mentioned in the docs but it's not asked to create yaml setup together with a config flow for new integrations

gjohansson-ST commented on homeassistant/components/rki_covid/sensor.py:
```diff
@@ -0,0 +1,207 @@
+    if coordinator is None or coordinator.data is None:
+        raise PlatformNotReady("Data coordinator could not be initialized!")
+
+    try:
+        district = config_entry.data[ATTR_COUNTY]
+        sensors = [
+            RKICovidNumbersSensor(coordinator, district, info_type)
+            for info_type in SENSORS
+        ]
+        async_add_entities(sensors, update_before_add=True)
+    except KeyError:
+        _LOGGER.error("Could not determine %s from config!", ATTR_COUNTY)
```
Why is a try needed here

thebino replied:
If an invalid/old config layout is used the integration setup will fail. (Integration was previous available via HACS)

gjohansson-ST replied:
Why not check the district then compared to some list and do things depending on outcome. This try block seems overcomplicated for the job.

thebino replied:
done

gjohansson-ST commented on homeassistant/components/rki_covid/sensor.py:
```diff
@@ -0,0 +1,207 @@
+            and self.district in self.coordinator.data
+        )
+
+    @property
+    def state(self) -> Optional[str]:
+        """Return current state."""
+        try:
+            return getattr(self.coordinator.data[self.district], self.info_type)
+        except AttributeError:
+            return None
```
Don't use state, use native value

thebino replied:
Where can I find details in the docs?

gjohansson-ST replied:
https://github.com/home-assistant/core/blob/dev/homeassistant/components/sensor/__init__.py#L260

thebino replied:
It is recommended in the official sample and necessary to publish the sensor data
https://github.com/home-assistant/example-custom-config/blob/master/custom_components/example_sensor/sensor.py#L34

gjohansson-ST replied:
Okey, so those examples needs to be fixed. If you see my link it implies not to use state, same with unit_of_measure.

thebino replied:
done

gjohansson-ST replied:
The example on your link above is now fixed according to the new guidelines which might help.

thebino replied:
done

gjohansson-ST commented on homeassistant/components/rki_covid/sensor.py:
```diff
@@ -0,0 +1,207 @@
+
+        data = coordinator.data[district]
+
+        if data.county:
+            self.name = f"{data.county} {info_type}"
+        else:
+            self.name = f"{data.name} {info_type}"
+        self.unique_id = f"{district}-{info_type}"
```
Use _attr for these and all the properties below
https://github.com/home-assistant/core/blob/dev/homeassistant/components/sensor/__init__.py#L251

thebino replied:
done

thebino replied:
There is no `_attr` property for any of them since they are only used by this custom sensor.

gjohansson-ST commented on homeassistant/components/rki_covid/sensor.py:
```diff
@@ -0,0 +1,207 @@
+        return STATE_CLASS_MEASUREMENT
+
+    @property
+    def unit_of_measurement(self):
```
Don't use `unit of measure`, use `native_unit_of_measure`

thebino replied:
I thought the `native_unit_of_measurement` should only be used if the unit will be modified by the SensorEntity
https://developers.home-assistant.io/blog/2021/08/12/sensor_temperature_conversion?_highlight=unit_of_measurement#native_unit_of_measurement

gjohansson-ST replied:
No, should always be used
https://github.com/home-assistant/core/blob/dev/homeassistant/components/sensor/__init__.py#L261

thebino replied:
This is also recommended in the official sample.
https://github.com/home-assistant/example-custom-config/blob/master/custom_components/example_sensor/sensor.py#L39

thebino replied:
done

gjohansson-ST commented on homeassistant/components/rki_covid/sensor.py:
```diff
@@ -0,0 +1,207 @@
+        else:
+            return "cases"
+
+    @property
+    def extra_state_attributes(self):
+        """Return device attributes."""
+        return {
+            ATTR_ATTRIBUTION: f"last updated {self.updated.strftime('%d %b, %Y  %H:%M:%S')} \n{ATTRIBUTION}"
+        }
```
```suggestion
```
Attribution should be set with `_attr_attribution` and not be in `extra_state_attributes`

thebino replied:
For platforms it is necessary to use the extra_state_attributes as described here:
https://developers.home-assistant.io/docs/dev_101_states?_highlight=extra_state_attributes#using-states-in-your-component

gjohansson-ST replied:
Atttribution should be your test that the data is collected from RKI
I suppose this `self.updated` contains a different time than when the sensor updated and that's why you want it as extra attributes?
Use another key in the dict and just have the datetime as the value in utc time and isoformatted

thebino replied:
I agree, changed to `_attr_attribution` during sensor initialization

gjohansson-ST reviewed (changes requested):
Here is some comments to get you started

thebino committed:
Update homeassistant/components/rki_covid/__init__.py

simplify unload

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

thebino committed:
Update homeassistant/components/rki_covid/__init__.py

simplify unload

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

thebino committed:
Update homeassistant/components/rki_covid/config_flow.py

remove unused CONNECTION_CLASS

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

thebino committed:
Update homeassistant/components/rki_covid/config_flow.py

remove unused CONNECTION_CLASS

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

thebino committed:
Update homeassistant/components/rki_covid/__init__.py

simplify platform setup

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

thebino committed:
Update homeassistant/components/rki_covid/__init__.py

simplify platform setup

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

thebino committed:
use internal attribute names

thebino committed:
use internal attribute names

thebino committed:
use platform enum

thebino committed:
use platform enum

thebino committed:
Revert "use platform enum"

This reverts commit c85e6482d8954fad6e1161ca1df7bffc20fa85ae.

thebino committed:
Revert "use platform enum"

This reverts commit c85e6482d8954fad6e1161ca1df7bffc20fa85ae.

thebino committed:
Revert "use internal attribute names"

This reverts commit 727fbcc357b0d0c0ef74ea45d50bd7394f186aa9.

thebino committed:
Revert "use internal attribute names"

This reverts commit 727fbcc357b0d0c0ef74ea45d50bd7394f186aa9.

thebino committed:
remove unused async_setup

thebino committed:
remove unused async_setup

thebino committed:
simplify sensor entry setup

thebino committed:
simplify sensor entry setup

thebino committed:
use internal attribution for last-update timestamp

thebino committed:
use internal attribution for last-update timestamp

thebino committed:
use internal attribution instead of property functions

thebino committed:
use internal attribution instead of property functions

thebino committed:
split sensors to make hospitalication exclusive for states

thebino committed:
split sensors to make hospitalication exclusive for states

iMicknl commented on homeassistant/components/rki_covid/sensor.py:
```diff
@@ -0,0 +1,212 @@
+    session = async_get_clientsession(hass)
+
+    if CONF_BASEURL in config:
+        _LOGGER.warning("Baseurl is not supported anymore.")
```
Why do you add this? This is a new integration, thus it doesn't make sense to deprecate a feature that never was supported.

thebino replied:
The integration was published via HACS before. In an earlier version there was support for a custom base url

iMicknl replied:
Why do you use `async_setup_platform` by the way, and not `async_setup_entry`? It seems you didn't fully migrate to a Config Flow with Config Entries.

iMicknl replied:
As @gjohansson-ST mentioned in the other comment, I would remove this part since it is not required and should not be added.

+ remove the other YAML parts / deprecation notice

thebino replied:
I removed all deprecated stuff and support for setup from config file

iMicknl commented on homeassistant/components/rki_covid/__init__.py:
```diff
@@ -0,0 +1,147 @@
+async def get_coordinator(hass: HomeAssistant, parser: RkiCovidParser):
+    """Get the data update coordinator."""
+    _LOGGER.debug("initialize the data coordinator.")
+    if DOMAIN in hass.data:
+        return hass.data[DOMAIN]
```
What is the purpose of this? I don't understand why you need this get_coordinator function.

Why not just instantiate your class that inherits from DataUpdateCoordinator in `async_setup_entry` and add it to your domain data? This way you can use it in your integration, instead of loading it in this way.

```python3
  coordinator = RKICovidDataUpdateCoordinator(hass)
  hass.data.setdefault(DOMAIN, {})[entry.entry_id] = {
        "coordinator": coordinator,
    }
```

thebino replied:
Haven't seen this implementation yet. Followed the integration guide in the developer docs.

iMicknl replied:
Can you link to that doc? I haven't seen this implementation yet.

thebino replied:
Here you go: https://developers.home-assistant.io/docs/integration_fetching_data#coordinated-single-api-poll-for-data-for-all-entities

thebino replied:
@iMicknl I've done some re-write of the integration. Still need to adjust some of the tests but would be great to get some additional feedback. Especially on the mentiponed `hass.data` and how I can re-use already loaded data in the sensor and config flow once loaded via the coordinator.

iMicknl commented on homeassistant/components/rki_covid/manifest.json:
```diff
@@ -0,0 +1,11 @@
+  "dependencies": [],
+  "codeowners": ["@thebino"],
+  "requirements": ["rki-covid-parser==1.3.1"],
+  "quality_scale": "gold",
```
I am not sure if you can list `gold` already. Have a look at https://developers.home-assistant.io/docs/integration_quality_scale_index/#gold-.
For example, I miss 
> Tests for fetching data from the integration and controlling it

thebino replied:
There is now a 91% coverage. I think this box is checked now as well again.

iMicknl commented on homeassistant/components/rki_covid/sensor.py:
```diff
@@ -0,0 +1,212 @@
+DISTRICT_SCHEMA = vol.Schema({vol.Required(CONF_NAME): cv.string})
+
+# schema for each platform sensor
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
```
Remove. Could you perhaps try to remove all unrelated / YAML setup parts? This will make reviewing a lot easier and faster.

thebino replied:
done

iMicknl commented on homeassistant/components/rki_covid/sensor.py:
```diff
@@ -0,0 +1,212 @@
+            self.name = f"{data.county} {info_type}"
+        else:
+            self.name = f"{data.name} {info_type}"
+        self.unique_id = f"{district}-{info_type}"
```
This is what @gjohansson-ST meant;

```suggestion
        self._attr_unique_id = f"{district}-{info_type}"
```

However, I am not sure if this is unique enough. See https://developers.home-assistant.io/docs/entity_registry_index/#unique-id-requirements.

gjohansson-ST replied:
Looks ok from first glance since I guess district here is county in the config flow which is set as unique id.
But he needs to clean this up before it's possible to review it properly

thebino replied:
done

iMicknl commented on homeassistant/components/rki_covid/sensor.py:
```diff
@@ -0,0 +1,212 @@
+class RKICovidNumbersSensor(CoordinatorEntity):
+    """Representation of a sensor."""
+
+    name = None
+    unique_id = None
```
```suggestion

```

thebino replied:
done

iMicknl commented on homeassistant/components/rki_covid/sensor.py:
```diff
@@ -0,0 +1,212 @@
+
+        data = coordinator.data[district]
+
+        if data.county:
+            self.name = f"{data.county} {info_type}"
+        else:
+            self.name = f"{data.name} {info_type}"
```
```suggestion
        if data.county:
            self._attr_name = f"{data.county} {info_type}"
        else:
            self._attr_name = f"{data.name} {info_type}"
```

thebino replied:
done

iMicknl commented on homeassistant/components/rki_covid/sensor.py:
```diff
@@ -0,0 +1,212 @@
+    @property
+    def icon(self):
+        """Return the icon."""
+        return {**SENSORS, **DISTRICT_SENSORS}[self.info_type]
```
Remove this and just set `self._attr_icon` in constructor.

thebino replied:
done

iMicknl commented on homeassistant/components/rki_covid/sensor.py:
```diff
@@ -0,0 +1,212 @@
+        return {**SENSORS, **DISTRICT_SENSORS}[self.info_type]
+
+    @property
+    def state_class(self):
```
Remove and just use attr.

thebino replied:
done

iMicknl commented on homeassistant/components/rki_covid/__init__.py:
```diff
@@ -0,0 +1,147 @@
+                for d in parser.districts:
+                    district = parser.districts[d]
+
+                    items[district.county] = DistrictData(
```
It seems that you are the author of `rki_covid_parser`. Why don't you directly add this typing in your python library?

thebino replied:
That is correct, the parser is only for downloading and parsing the data.
Since merging is a frontend requirement, it should not be part of a parsing library

thebino committed:
use platform enum - again

thebino committed:
use platform enum - again

thebino committed:
using union type expressions

thebino committed:
using union type expressions

thebino committed:
remove support for file configurations

thebino committed:
remove support for file configurations

thebino committed:
replace icon property with _attr_icon

thebino committed:
replace icon property with _attr_icon

thebino committed:
extract coordinator and add tests

thebino committed:
extract coordinator and add tests

thebino committed:
adjust tests

thebino committed:
adjust tests

thebino committed:
sort inside config flow

thebino committed:
sort inside config flow

thebino committed:
add more tests

thebino committed:
add more tests

thebino committed:
clean-up

thebino committed:
clean-up

thebino commented:
@gjohansson-ST Would be great to have another round of review. So we can finish and merge this integration soon. Thanks already for all your hints and ideas.

frenck commented on homeassistant/components/rki_covid/__init__.py:
```diff
@@ -0,0 +1,59 @@
+"""RKI Covid numbers integration."""
+import logging
+
+from homeassistant.components.rki_covid.coordinator import RkiCovidDataUpdateCoordinator
```
```suggestion
from .coordinator import RkiCovidDataUpdateCoordinator
```

frenck commented on homeassistant/components/rki_covid/__init__.py:
```diff
@@ -0,0 +1,59 @@
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up the RKI covid numbers integration from a config entry."""
+    _LOGGER.debug("async_setup_entry from a config entry")
+
```
I would not consider this a really helpful or required debug log.

```suggestion
```

frenck commented on homeassistant/components/rki_covid/__init__.py:
```diff
@@ -0,0 +1,59 @@
+    # trigger initial refresh
+    await coordinator.async_config_entry_first_refresh()
+
+    # put data into
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN].setdefault(DATA_CONFIG_ENTRY, {})
+    hass.data[DOMAIN][DATA_CONFIG_ENTRY][entry.entry_id] = coordinator
```
We don't need an additional dictionary here, as it seems there is no additional data put in.

```suggestion
    # put data into
    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = coordinator
```

frenck commented on homeassistant/components/rki_covid/__init__.py:
```diff
@@ -0,0 +1,59 @@
+    # set default options if not already set
+    if not entry.options:
+        options = {
+            CONF_SCAN_INTERVAL: entry.data.get(
+                CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL
+            ),
+        }
```
The scan interval should not be configurable via the integration options. If one wants to use a different update interval, they can achieve that using the `homeassistant.update_entity` service in an automation.

Please remove this feature.

frenck commented on homeassistant/components/rki_covid/__init__.py:
```diff
@@ -0,0 +1,59 @@
+        hass.config_entries.async_update_entry(entry, options=options)
+
+    # setup platform
+    hass.config_entries.async_setup_platforms(entry, [Platform.SENSOR])
```
Please create a constant `PLATFORMS` at the top of the file with this list.

```suggestion
    hass.config_entries.async_setup_platforms(entry, PLATFORMS)
```

frenck commented on homeassistant/components/rki_covid/__init__.py:
```diff
@@ -0,0 +1,59 @@
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    _LOGGER.debug("async_unload_entry %s", entry.data)
```
```suggestion
```

frenck commented on homeassistant/components/rki_covid/__init__.py:
```diff
@@ -0,0 +1,59 @@
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    _LOGGER.debug("async_unload_entry %s", entry.data)
+    unload_ok = await hass.config_entries.async_unload_platforms(
+        entry, [Platform.SENSOR]
+    )
+
+    if unload_ok:
+        hass.data[DOMAIN][DATA_CONFIG_ENTRY].pop(entry.entry_id)
+        if not hass.data[DOMAIN][DATA_CONFIG_ENTRY]:
+            hass.data[DOMAIN].pop(DATA_CONFIG_ENTRY)
+
```
```suggestion
    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
        hass.data[DOMAIN].pop(entry.entry_id)
```

frenck commented on homeassistant/components/rki_covid/config_flow.py:
```diff
@@ -0,0 +1,55 @@
+from typing import Any
+import voluptuous as vol
+from homeassistant import config_entries
+from homeassistant.components.rki_covid.coordinator import RkiCovidDataUpdateCoordinator
```
```suggestion
from .coordinator import RkiCovidDataUpdateCoordinator
```

frenck commented on homeassistant/components/rki_covid/config_flow.py:
```diff
@@ -0,0 +1,55 @@
+
+    _options: dict[str, Any] | None = None
+
+    async def async_step_user(self, user_input: dict[str, Any] | None = None):
```
Missing return type

```suggestion
    async def async_step_user(self, user_input: dict[str, Any] | None = None) -> FlowResult:
```

frenck commented on homeassistant/components/rki_covid/config_flow.py:
```diff
@@ -0,0 +1,55 @@
+        if self._options is None:
+            self._options = {}
+
+            # add items from coordinator
+            coordinator = RkiCovidDataUpdateCoordinator(self.hass)
+            await coordinator.async_config_entry_first_refresh()
```
Lets not use the coordinator here, instead, use the client library.

thebino replied:
The whole accumulation of the data is done in the coordinator since it has nothing to do with the client library. Wouldn't it be good to keep the coordinator to decrease code duplitaces?

frenck replied:
> Wouldn't it be good to keep the coordinator to decrease code duplitaces?

It isn't needed to use the coordinator for that, instead, abstract the accumulation in a separate method that is used by the coordinator and the config flow.

thebino replied:
Thats true, I extracted it and use the parser directly for the config_flow

frenck commented on homeassistant/components/rki_covid/config_flow.py:
```diff
@@ -0,0 +1,55 @@
+            coordinator = RkiCovidDataUpdateCoordinator(self.hass)
+            await coordinator.async_config_entry_first_refresh()
+
+            _LOGGER.debug(f"Coordinator: {coordinator}")
+
```
No need to dump the full coordinator object...

frenck commented on homeassistant/components/rki_covid/config_flow.py:
```diff
@@ -0,0 +1,55 @@
+            self._abort_if_unique_id_configured()
+
+            # User is done adding sensors, create the config entry.
+            _LOGGER.debug("Create entry from Configuration UI")
```
```suggestion
```

frenck commented on homeassistant/components/rki_covid/coordinator.py:
```diff
@@ -0,0 +1,141 @@
+
+    parser: RkiCovidParser
+
+    def __init__(self, hass: HomeAssistant):
```
```suggestion
    def __init__(self, hass: HomeAssistant) -> None:
```

thebino committed:
clean-up

thebino committed:
clean-up

thebino committed:
extract data accumulation from coordinator

thebino committed:
extract data accumulation from coordinator

thebino committed:
isort

thebino committed:
isort

thebino committed:
adjust ci

thebino committed:
adjust ci

thebino committed:
fix tests

thebino committed:
fix tests

thebino committed:
adjust translations

thebino committed:
adjust translations

gjohansson-ST commented on homeassistant/components/rki_covid/sensor.py:
```diff
@@ -0,0 +1,172 @@
+}
+
+
+async def async_setup_platform(
+    hass: HomeAssistantType,
+    config: ConfigType,
+    async_add_entities: AddEntitiesCallback,
+    discovery_info: DiscoveryInfoType | None = None,
+) -> None:
+    """Set up the sensor platform."""
+    coordinator = RkiCovidDataUpdateCoordinator(hass)
+    await coordinator.async_config_entry_first_refresh()
+
+    if coordinator is None or coordinator.data is None:
+        raise PlatformNotReady("Data coordinator could not be initialized!")
+
+    districts = config[CONF_DISTRICTS]
+
+    # district sensors
+    sensors = [
+        RKICovidNumbersSensor(coordinator, district[CONF_DISTRICT_NAME], info_type)
+        for info_type in SENSORS
+        for district in districts
+    ]
+    async_add_entities(sensors, update_before_add=True)
+
+    # sensors on state level (additional sensors)
+    for district in districts:
+        if district.startswith("BL"):
+            district_sensors = [
+                RKICovidNumbersSensor(coordinator, district, info_type)
+                for info_type in DISTRICT_SENSORS
+            ]
+            async_add_entities(district_sensors, update_before_add=True)
```
Should be removed as it's setup from entry only (config flow)

thebino replied:
Can you point me what exactly I need to change? In my understanding this integration should use the platform concept since it is containing multiple sensors for a single unit.
Can't figure out what I have to change here.

gjohansson-ST replied:
async_setup_platform is used when you load the platform from config.yaml. Here you use a config flow so it will use async_setup_entry

gjohansson-ST commented on homeassistant/components/rki_covid/sensor.py:
```diff
@@ -0,0 +1,172 @@
+    _LOGGER.debug("create sensor from config entry %s", config_entry.data)
+    coordinator = hass.data[DOMAIN][config_entry.entry_id]
+
+    if coordinator is None or coordinator.data is None:
+        raise PlatformNotReady("Data coordinator could not be initialized!")
```
Not needed as it would not come this far if coordinator or data not available

thebino replied:
done

gjohansson-ST commented on homeassistant/components/rki_covid/sensor.py:
```diff
@@ -0,0 +1,172 @@
+
+        self.district = district
+        self.info_type = info_type
+        self.updated = datetime.now()
```
Not used?

thebino replied:
done.

gjohansson-ST commented on homeassistant/components/rki_covid/sensor.py:
```diff
@@ -0,0 +1,172 @@
+    @property
+    def available(self) -> bool:
+        """Return True if entity is available."""
+        return (
+            self.coordinator.last_update_success
+            and self.district in self.coordinator.data
+        )
```
```suggestion
        return (
            super().available
            and self.district in self.coordinator.data
        )
```

thebino replied:
done

gjohansson-ST reviewed (changes requested):
Some additional comments

thebino committed:
remove unused code and update available check

thebino committed:
remove unused code and update available check

cgtobi commented:
@thebino please rebase your PR onto dev rather than merging it. `git pull --rebase upstream dev` usually does the job for me.

thebino commented:
> @thebino please rebase your PR onto dev rather than merging it. `git pull --rebase upstream dev` usually does the job for me.

sure, sorry for that.

balloob committed:
Add helper for media players to handle HA hosted media (#66120)

* Sonos to sign all HASS urls

* Don't sign if queries in url

* Extract media player hass URL handling to helper

frenck committed:
Extract Spotify media browsing into a module (#66175)

epenet committed:
Adjust coverage to include all config flows (#66193)

* Adjust tradfri

* Adjust huawei_lte

* Adjust iqvia

* Adjust wiffi

* Adjust solarlog

* Adjust lifx

* Adjust doorbird

* Adjust rachio

* Adjust starline

* Adjust konnected

* Adjust ambient_station

* Adjust tado

* Adjust point

* Adjust daikin

* Adjust hangouts

* Adjust ifttt

* Adjust ios

* Adjust life360

* Adjust sms

* Adjust spider

* Adjust upnp

* Adjust hassfest

epenet committed:
Move Freebox reboot service to a button entity (#65501)

* Add restart button to freebox

* Add warning

* restart => reboot

* Add button tests

Co-authored-by: epenet <epenet@users.noreply.github.com>

eifinger committed:
add @eifinger as google_travel_time_codeowner (#66215)

eifinger committed:
add @eifinger as waze_travel_time codeowner (#66214)

jjlawren committed:
Schedule activity checks when using manual hosts (#65970)

iMicknl committed:
Add siren platform to Overkiz (#65300)

Commit:
[ci skip] Translation update

jjlawren committed:
Sonos lock subscription actions (#66204)

* Move exception logging to helper

* Wrap subscription in lock

* Rewrite subscribe method to use new logging helper

* Mark entitites as unavailable sooner to avoid unnecessary polls

* Rename unclear method and update docstring

* Move lock to unsubscribe

ggravlingen committed:
Fix tradfri device name (#66219)

j-a-n committed:
Add Moehlenhoff Alpha2 underfloor heating system integration (#42771)

* Add Moehlenhoff Alpha2 underfloor heating system integration

* isort changes

* flake8 changes

* Do not exclude config_flow.py

* pylint changes

* Add config_flow test

* correct requirements_test_all.txt

* more tests

* Update test description

* Test connection and catch TimeoutError in async_setup_entry

* Add version to manifest file

* Remove version from manifest file

* Replace tests.async_mock.patch by unittest.mock.patch

* Update moehlenhoff-alpha2 to version 1.0.1

* Update requirements for moehlenhoff-alpha2 1.0.1

* Update moehlenhoff-alpha2 to 1.0.2

* Use async_setup_platforms

* Use async_unload_platforms

* Separate connection and devices for each entry_id

* Use async_track_time_interval to schedule updates

* Check if input is valid before checking uniqueness

* Move Exception handling to validate_input

* Catch aiohttp.client_exceptions.ClientConnectorError

* Remove translation files

* Mock TimeoutError

* Fix data update

* Replace current callback implementation with ha dispatcher

* Return False in should_poll

* Remove unused argument

* Remove CONNECTION_CLASS

* Use _async_current_entries

* Call async_schedule_update_ha_state after data update

* Remove unneeded async_setup

Co-authored-by: Milan Meulemans <milan.meulemans@live.be>

* Remove unneeded async_setup_platform

Co-authored-by: Milan Meulemans <milan.meulemans@live.be>

* Set Schema attribute host required

Co-authored-by: Milan Meulemans <milan.meulemans@live.be>

* Remove unused Exception class

Co-authored-by: Milan Meulemans <milan.meulemans@live.be>

* Update manifest.json

Co-authored-by: Milan Meulemans <milan.meulemans@live.be>

* pylint constructor return type None

* Replace properties by class variables

* use pass instead of return

* Remove unused sync update method

* remove property hvac_action

* remove pass

* rework exception handling

* Update homeassistant/components/moehlenhoff_alpha2/config_flow.py

Co-authored-by: Milan Meulemans <milan.meulemans@live.be>

* Correct indentation

* catch Exception in validate_input

* Replace HomeAssistantType with HomeAssistant

* Update to moehlenhoff-alpha2 1.0.3

* Allow to switch between heating and cooling mode

* Update moehlenhoff-alpha2 to version 1.0.4

* Update heatarea data after setting target temperature

* Support hvac_action

* Fix heatarea update with multiple bases

* Update data after setting preset mode

* Use custom preset modes like defined by device

* Fix config flow test

* Fix test_duplicate_error

* Rename property to extra_state_attributes

Rename property device_state_attributes to extra_state_attributes and
return lowercase keys in dict.

* Refactor using DataUpdateCoordinator

* Remove _attr_should_poll

* Raise HomeAssistantError on communication error

Catch HTTPError instead of broad except and reraise as HomeAssistantError

* Change DataUpdateCoordinator name to alpha2_base

* Refresh coordinator before setting data

* Raise ValueError on invalid heat area mode

* Rename heatarea to heat_area

* Set type annotation in class attribute

* Move coordinator to top

* Move exception handling to the coordinator

* Use heat_area_id directly

* Sore get_cooling() result into local var

* Add explanation of status attributes

and remove BLOCK_HC

* Fix pylint warnings

* from __future__ import annotations

* Use Platform Enum

* Move data handling to coordinator

* Remove property extra_state_attributes

* Add missing annotations

* Update moehlenhoff-alpha2 to version 1.1.2

* Rework tests based on the scaffold template

* Set also heat/cool/day/night temp with target temp

* Remove unneeded code from tests

Co-authored-by: Milan Meulemans <milan.meulemans@live.be>

milanmeu committed:
Bump aioaseko to 0.0.2 to fix issue (#66240)

mib1185 committed:
bump py-synologydsm-api to 1.0.6 (#66226)

raman325 committed:
Simplify get_unique_id helper function for zwave_js (#66221)

Danielhiversen committed:
Add more sensors for users with Tibber Pulse (#66201)

* Tibber, Add stats and sensors for homes with real time meter

Signed-off-by: Daniel Hjelseth Høyer <github@dahoiv.net>

* Tibber stats

Signed-off-by: Daniel Hjelseth Høyer <github@dahoiv.net>

* Monthly peak hour

balloob committed:
Sort media sources (#66237)

* Sort media sources

* Flatten media sources and sort Cast children

* Only expand if possible

epenet committed:
Enable basic type checking in demo init (#66218)

* Enable basic type checking in demo init

* Remove from mypy ignore list

jbouwh committed:
Mqtt move to .const (#65631)

frenck committed:
Refactor Plugwise command handling (#66202)

epenet committed:
Enable basic type checking in demo platforms (#66212)

* Adjust type hints in demo platforms

* Adjust mypy config

* Adjust name

frenck committed:
More cleanup in Plugwise switch (#66254)

epenet committed:
Enable basic type checking in template (#66222)

* Fix binary_sensor

* Adjust button

* Adjust fan

* Adjust select

* Adjust template_entity

* Adjust trigger_entity

* Adjust weather

* Adjust init

* Adjust number

* Adjust None check

emontnemery committed:
Tweak constant config_entries.DISCOVERY_SOURCES (#66249)

frenck committed:
More cleanup in Plugwise binary sensor (#66255)

frenck committed:
Add additional switches to Plugwise (#66261)

frenck committed:
Remove Plugwise Auxiliary sensors (#66259)

bdraco committed:
Add RGBW/RGBWW support to WiZ (#66196)

balloob committed:
Allow uploading media to media folder (#66143)

jarodwilson committed:
Add unique_id for decora_wifi lights (#66142)

I have some decora_wifi switches that I want to be able to move around
to different rooms within Home Assistant, and for that to be practical,
they need unique IDs, so here we are.

Signed-off-by: Jarod Wilson <jarodwilson@gmail.com>

frenck committed:
More cleanup in Plugwise climate (#66257)

frenck committed:
More cleanup in Plugwise sensor (#66274)

frenck committed:
Add Flame State binary sensor to Plugwise (#66275)

frenck committed:
Plugwise HVAC/Preset mode fixes (#66273)

bachya committed:
Clean up unnecessary branch in SimpliSafe (#66268)

frenck committed:
Cleanup existing Plugwise tests and test fixtures (#66282)

* Cleanup existing Plugwise tests and test fixtures

* More cleanup

frenck committed:
Extend Plugwise climate support (#66278)

frenck committed:
Add humidity sensor to Plugwise (#66280)

emontnemery committed:
Drop MQTT import flow (#66160)

* Drop MQTT import flow

* Reload manually configured MQTT entities when config entry is setup

* Address review comments

* Actually remove the import flow

jjlawren committed:
Tweak Sonos activity monitoring (#66207)

* Tweak Sonos activity monitoring

* Support new 'sleeping' vanish reason
* Check availability before last-ditch check
* Use short-timeout call for last-ditch check
* Adjust reboot logging message and severity

* Simplify activity check failure

balloob committed:
Fix shutil import for local source (#66286)

jjlawren committed:
Handle more Sonos favorites in media browser (#66205)

uvjustin committed:
Catch ConnectionResetError when writing MJPEG in camera (#66245)

bdraco committed:
Add discovery on network up to WiZ (#66144)

frenck committed:
Type Spotify hass data (#66285)

jjlawren committed:
Skip polling Sonos audio input sensor when idle (#66271)

elupus committed:
Correct philips_js usage of the overloaded coordinator (#66287)

elupus committed:
Add diagnostics for philips_js (#66233)

* Add diagnostics for philips_js

* Update homeassistant/components/philips_js/diagnostics.py

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

* Update homeassistant/components/philips_js/diagnostics.py

Co-authored-by: Robert Svensson <Kane610@users.noreply.github.com>

* Also redact username/password

They are really not that secret, but seem logical.

* Redact unique id

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>
Co-authored-by: Robert Svensson <Kane610@users.noreply.github.com>

bdraco committed:
Guard against 0 value for color temp in WiZ when turning off (#66295)

Commit:
[ci skip] Translation update

allenporter committed:
Bump google-nest-sdm to 1.7.1 (minor patch) (#66304)

bdraco committed:
Fix august token refresh when data contains characters outside of latin1 (#66303)

* WIP

* bump version

* bump

dingusdk committed:
Prepare for Ihc config flow (#64852)

* Extracting group and extra  info from ihc products

* Make suggested area not optional

* Revert back to assignment expression :=

* Make auto setup show device info for all platforms

* Change code comment to #

* Add return typing

* Remove device_info key without value

* get_manual_configuration typings for everything

* Adding IHCController typings

* Remove "ihc" from unique id

* Remove device_info

* Separator in unique id

* Return typing on ihc_setup

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

* Update homeassistant/components/ihc/ihcdevice.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

* Update homeassistant/components/ihc/ihcdevice.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

* Raise ValueError instead of logging an error

* Update homeassistant/components/ihc/service_functions.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

* Catch up with dev

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

emontnemery committed:
Create MQTT discovery flow when manual config is present (#66248)

* Create MQTT discovery flow when manual config is present

* Change to integration_discovery flow

* Add test

* Add default handler for integration_discovery

ufodone committed:
Disable zone bypass switch feature (#66243)

* Add configuration option to disable the creation of zone bypass switches

* Removed temporary workaround and bumped pyenvisalink version to pick up the correct fix.

* Remove zone bypass configuration option and disable zone bypass switches per code review instructions.

stebet committed:
Add Icelandic to list of supported Azure languages (#66310)

It was missing, as it is a supported language as per the list here: https://docs.microsoft.com/en-us/azure/cognitive-services/speech-service/language-support#text-to-speech

frenck committed:
Add heating and cooling binary sensors to Plugwise (#66317)

cereal2nd committed:
Bump velbusaio to 2022.2.4 (#66321)

ludeeus committed:
Add guard for invalid EntityCategory value (#66316)

thecode committed:
Fix webostv restored supported features turn on (#66318)

* Fix webostv restored supported features turn on

* Remove ternary operator expression

starkillerOG committed:
bump motionblinds to 0.5.12 (#66323)

bdraco committed:
Fix WiZ bulb being offline if kelvin limits cannot be obtained (#66305)

frenck committed:
Don't requests known Spotify playlist (#66313)

frenck committed:
Use DataUpdateCoordinator for Spotify devices (#66314)

bdraco committed:
Reduce number of parallel api calls to august (#66328)

epenet committed:
Fix raspihats initialization (#66330)

Co-authored-by: epenet <epenet@users.noreply.github.com>

frenck committed:
Fix PVOutput when no data is available (#66338)

frenck committed:
Fix CPUSpeed with missing info (#66339)

Jc2k committed:
Prepare for new aiohomekit lifecycle API (#66340)

stuart-c committed:
Upgrade OVO library to v1.2.0 (#66210)

allenporter committed:
Fix nest streams that get stuck broken (#66334)

bdraco committed:
Add unique id to lutron caseta config entry when missing (#66346)

bdraco committed:
Add additional OUI for G3 wifi cameras to unifiprotect (#66349)

bdraco committed:
Add additional oui to blink (#66348)

bdraco committed:
Add dhcp discovery to Sensibo for non-HomeKit devices (#66350)

Kirmas committed:
Initial xiaomi_miio support for dmaker.airfresh.a1/t2017 (#66331)

* Initial support for dmaker.airfresh.a1/t2017

* fix typo

roysjosh committed:
Get discovered zeroconf IPv6 addresses (#65462)

PanicRide committed:
New amcrest binary sensor to monitor doorbell button (#66302)

* New binary sensor to monitor doorbell button

* New binary sensor to monitor doorbell button

bdraco committed:
Add button to wake august locks from deep sleep (#66343)

bdraco committed:
Fix zwave_me zeroconf mocking (#66356)

jeeftor committed:
Add a base class for Intellifire entities (#65077)

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

gjohansson-ST committed:
Code cleanup yale_smart_alarm (#65081)

Sander0542 committed:
Add device info to Solax (#65244)

GitHub-Action committed:
[ci skip] Translation update

bdraco committed:
Fix uncaught exception during WiZ discovery during firmware update (#66358)

ctalkington committed:
Add select platform to roku (#66133)

allenporter committed:
Publish Nest Motion/Person events with optional user defined zone information (#66187)

Publish Nest events with zone information if present. User defined zones are configured
in the Google Home app, and are published with Motion/Person event.

gjohansson-ST committed:
Fix supported features sensibo (#65895)

* Fix features not available

* Partial revert

* Apply active features

* Fix from review

Bre77 committed:
Add Diagnostics (#65755)

Jc2k committed:
Simplify the homekit_controller unignore journey (#66353)

Jc2k committed:
Stop homekit_controller using backend specific API's (#66375)

gjohansson-ST committed:
Improve code quality sql (#65321)

gjohansson-ST committed:
Improve code quality telnet (#65239)

gjohansson-ST committed:
Improve code quality command_line (#65333)

gjohansson-ST committed:
Bump yalesmartalarmclient to v0.3.8 (#66381)

davet2001 committed:
Fix missing refactors of EntityCategory.XXX (#66379)

* Fix missing refactors of EntityCategory.XXX

* Fix entity_category refactor for homewizard

gjohansson-ST committed:
Code quality scrape (#65441)

gjohansson-ST committed:
Implement diagnostics for yale_smart_alarm (#65085)

roysjosh committed:
Specify specific Nanoleaf models in the manifest (#66326)

doudz committed:
Update zigpy-zigate to 0.8.0 (#66289)

nklebedev committed:
Fix typo in ebusd WaterPressure const (#66355)

frenck committed:
Fix Spotify session token refresh (#66386)

corneyl committed:
Add Picnic re-auth flow (#62938)

* Add re-auth handler for Picnic

* Extracted authentication part so right form/errors can be shown during re-auth flow

* Add tests for Picnic's re-authentication flow

* Simplify re-auth flow by using the same step as step_user

* Use user step also for re-auth flow instead of having an authenticate step

* Add check for when re-auth is done with different account

* Remove unnecessary else in Picnic config flow

* Fix the step id in the translation strings file

cdce8p committed:
Fix typing [roku] (#66397)

gjohansson-ST committed:
Code quality file (#65258)

kpine committed:
Add is_controller_node flag to WS node status (#66404)

cdce8p committed:
Fix error decorator [sonos] (#66399)

chemelli74 committed:
Fix mesh role for Fritz old devices (#66369)

allenporter committed:
Redact stream url credentials in debug logging (#66407)

Kane610 committed:
Revert Amcrest change (#66412)

GitHub-Action committed:
[ci skip] Translation update

allenporter committed:
Improve stream robustness by always retrying worker (#66417)

Improve stream robustness by always retrying in the worker on failure, rather than only when
keepalive is enabled.

This will make cloud cameras like nest more robust, since they have a tendency to be flaky. This
is also needed to improve client side retry behavior because when the client attempts to retry,
the stream token is already revoked because the worker stopped.

The worker will still idle timeout if no streams are present, so it won't go on forever if no
frontend is viewing the stream.

Jc2k committed:
Replace use of deprecated APIs in aiohomekit (#66409)

thebino committed:
remove unused async_setup_platform

thebino committed:
Merge branch 'rki_covid' of github.com:thebino/core into rki_covid

homeassistant commented:
Hi @GitHub-Action, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#60506).

Once you do that we will be able to review and accept this pull request.

Thanks!
